src/codechicken/nei/recipe/GuiRecipe.java

```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
         yShift = handlerInfo == null ? 0 : handlerInfo.getYShift();
     }
 
-
     public void setRecipePage(int idx) {
+        setRecipePage(idx, 0);
+    }
+
+    public void setRecipePage(int idx, int idp) {
         recipetype = idx;
         if (recipetype < 0) recipetype = currenthandlers.size() - 1;
         else if (recipetype >= currenthandlers.size()) recipetype = 0;
 
         handler = currenthandlers.get(recipetype);
         handlerInfo = getHandlerInfo(handler);
-        page = 0;
+        page = Math.min(idp, (handler.numRecipes() - 1) / getRecipesPerPage());
         recipeTabs.calcPageNumber();
         checkYShift();
         initOverlayButtons();
     }
 
+    public void openTargetRecipe(String recipeId) 
+    {
+
+        if (recipeId == "") {
+            return;
+        }
+
+        String[] expected = recipeId.split("\\|", 2);
+        Integer recipetype = 0;
+        Integer page = 0;
+
+        for (int j = 0; j < currenthandlers.size(); j++) {
+            IRecipeHandler localHandler = currenthandlers.get(j);
```

Comment:
nit: final

Reply:
here I do not understand what you mean. Whee need `final`? in 227 line?